pkg/controller/supportbundlecollection/validate.go

```diff
@@ -0,0 +1,114 @@
+// Copyright 2022 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package supportbundlecollection
+
+import (
+	"encoding/json"
+	"fmt"
+	"reflect"
+
+	admv1 "k8s.io/api/admission/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/klog/v2"
+
+	crdv1alpha1 "antrea.io/antrea/pkg/apis/crd/v1alpha1"
+	"antrea.io/antrea/pkg/apiserver/storage"
+)
+
+type Validator struct {
+	store storage.Interface
+}
+
+func NewSupportBundleCollectionValidator(supportBundleCollectionStore storage.Interface) *Validator {
+	return &Validator{
+		store: supportBundleCollectionStore,
+	}
+}
+
+func (v *Validator) Validate(review *admv1.AdmissionReview) *admv1.AdmissionResponse {
+	var msg string
+	allowed := true
+
+	klog.V(2).Info("Validating SupportBundleCollection", "request", review.Request)
+	var newObj, oldObj crdv1alpha1.SupportBundleCollection
+	if review.Request.Object.Raw != nil {
+		if err := json.Unmarshal(review.Request.Object.Raw, &newObj); err != nil {
+			klog.ErrorS(err, "Error de-serializing current SupportBundleCollection")
+			return newAdmissionResponseForErr(err)
+		}
+	}
+	if review.Request.OldObject.Raw != nil {
+		if err := json.Unmarshal(review.Request.OldObject.Raw, &oldObj); err != nil {
+			klog.ErrorS(err, "Error de-serializing old IPPool")
+			return newAdmissionResponseForErr(err)
+		}
+	}
+
+	validateProcessingCollection := func(checkSpecUpdateOnly bool) *admv1.AdmissionResponse {
+		var msg string
+		allowed := true
+		_, exists, _ := v.store.Get(oldObj.Name)
+		if exists {
+			if checkSpecUpdateOnly {
+				allowed = reflect.DeepEqual(oldObj.Spec, newObj.Spec)
+			} else {
+				allowed = false
+			}
+			if !allowed {
+				msg = fmt.Sprintf("SupportBundleCollection %s is started, cannot be updated or deleted", oldObj.Name)
```

Comment:
Can there be a race condition with the controller worker?

Reply:
It is possible. I added a lock before accessing the store

Reply:
I do not mean accessing the store. I thought the store itself is thread safe, no?

Earlier I did not see you ignore spec changes in controller, and so was asking what will happen if there is a race condition that a spec change passes the check here. I now got we will ignore spec changes and I am fine.